hbase-balancer/src/main/java/org/apache/hadoop/hbase/CatalogAccessor.java

```diff
@@ -0,0 +1,1071 @@
+   * @param connection connection we're using
+   * @param visitor Visitor invoked against each row in tables family.
+   */
+  public static void fullScanTables(Connection connection,
+    final ClientMetaTableAccessor.Visitor visitor) throws IOException {
+    scanMeta(connection, null, null, QueryType.TABLE, visitor);
+  }
+
+  /**
+   * Performs a full scan of <code>hbase:meta</code>.
+   * @param connection connection we're using
+   * @param type scanned part of meta
+   * @return List of {@link Result}
+   */
+  private static List<Result> fullScan(Connection connection, QueryType type) throws IOException {
+    ClientMetaTableAccessor.CollectAllVisitor v = new ClientMetaTableAccessor.CollectAllVisitor();
+    scanMeta(connection, null, null, type, v);
+    return v.getResults();
+  }
+
+  /**
+   * Callers should call close on the returned {@link Table} instance.
+   * @param connection connection we're using to access Meta
+   * @return An {@link Table} for <code>hbase:meta</code>
+   * @throws NullPointerException if {@code connection} is {@code null}
+   */
+  public static Table getCatalogHTable(final Connection connection, TableName catalogTableName)
+    throws IOException {
+    if (!isCatalogTable(catalogTableName)) {
+      throw new IllegalStateException("Table supplied is not a catalog table: " + catalogTableName);
+    }
+    // We used to pass whole CatalogTracker in here, now we just pass in Connection
+    Objects.requireNonNull(connection, "Connection cannot be null");
+    if (connection.isClosed()) {
+      throw new IOException("connection is closed");
+    }
+    return connection.getTable(catalogTableName);
+  }
+
+  /**
+   * Gets the region info and assignment for the specified region.
+   * @param connection connection we're using
+   * @param regionName Region to lookup.
+   * @return Location and RegionInfo for <code>regionName</code>
+   * @deprecated use {@link #getRegionLocation(Connection, byte[])} instead
+   */
+  @Deprecated
+  public static Pair<RegionInfo, ServerName> getRegion(Connection connection, byte[] regionName)
+    throws IOException {
+    HRegionLocation location = getRegionLocation(connection, regionName);
+    return location == null ? null : new Pair<>(location.getRegion(), location.getServerName());
+  }
+
+  /**
+   * Returns the HRegionLocation from meta for the given region
+   * @param connection connection we're using
+   * @param regionName region we're looking for
+   * @return HRegionLocation for the given region
+   */
+  public static HRegionLocation getRegionLocation(Connection connection, byte[] regionName)
+    throws IOException {
+    byte[] row = regionName;
+    RegionInfo parsedInfo = null;
+    try {
+      parsedInfo = CatalogFamilyFormat.parseRegionInfoFromRegionName(regionName);
+      row = CatalogFamilyFormat.getMetaKeyForRegion(parsedInfo);
+    } catch (Exception parseEx) {
+      return null;
+      // Ignore. This is used with tableName passed as regionName.
```

Comment:
Does this happen? Should the javadoc allow this as a case? Is it normal that we get regioninfo or tablename?

Reply:
Yeah this did happen if I remember right. I'll update the javadoc then? Not sure I understand what you by is it normal?